### PR TITLE
Implement continuous batching

### LINKED ISSUE #1020: [FEATURE] Concurrent inference / continuous batching.
Hi,

During our testing  we find that EXO does not support concurrent inference by multiple users. Please help to clarify it. Thanks.

Nan

### PR DESCRIPTION
**Pull request by rltakashige**
## Motivation

Following the changes made in #1632 !
Closes #1020

### CODE REVIEW

**Comment by Evanev7** [comment]
ready for review? or not quite yet

File: src/exo/worker/engines/mlx/generator/batch_generate.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+                    f"KV cache hit: {prefix_hit_length}/{len(all_prompt_tokens)} tokens "
+                    f"cached ({100 * prefix_hit_length / len(all_prompt_tokens):.1f}%)"
+                )
+                prompt_tokens = remaining_tokens
+            else:
+                cache = make_kv_cache(self.model)
+        else:
+            cache = make_kv_cache(self.model)
+
+        seed = task_params.seed or 42
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/generator/batch_generate.py]: minor: should use explicit None check - this will override 0 with 42.

File: src/exo/worker/engines/mlx/generator/batch_generate.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+                all_prompt_tokens,
+                list(cache),
+                cache_snapshots,
+                prefix_hit_length,
+                matched_index,
+            )
+
+        last_tokens = prompt_tokens[-2:]
+
+        logits_processors: list[Any] = []
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/generator/batch_generate.py]: i believe we have some repetition penalty logits processors? assuming that merged, this should presumably duplicate that logic (or maybe a single make_logits_processors idk)

File: src/exo/worker/engines/mlx/generator/batch_generate.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+        if is_bench:
+            eos_ids = eos_ids_from_tokenizer(self.tokenizer)
+            logits_processors = [ban_token_ids(eos_ids)]
+
+        max_tokens = task_params.max_output_tokens or MAX_TOKENS
+
+        uids = self._mlx_gen.insert(
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/generator/batch_generate.py]: can this ever return multiple uids? we should guard that case

Comment (Evanev7) [reply src/exo/worker/engines/mlx/generator/batch_generate.py]: reading further this seems to be for multiple insertion - assuming we have no interest in multiple insertion, we should just assert it's a single uid.

File: src/exo/worker/engines/mlx/generator/batch_generate.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        )
+
+        return uids[0]
+
+    def step(self) -> list[tuple[int, GenerationResponse]]:
+        if not self.has_work:
+            return []
+
+        responses = self._mlx_gen.next()
+        mx.clear_cache()
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/generator/batch_generate.py]: this clear_cache could use a comment imo

File: src/exo/worker/engines/mlx/generator/batch_generate.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+        results: list[tuple[int, GenerationResponse]] = []
+
+        for response in responses:
+            if response.uid not in self._active_tasks:
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/generator/batch_generate.py:223]: feels like an error we should report

File: src/exo/worker/engines/mlx/generator/batch_generate.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+                    prompt_tokens=total_prompt_tokens,
+                    completion_tokens=state.completion_tokens,
+                    total_tokens=total_prompt_tokens + state.completion_tokens,
+                    prompt_tokens_details=PromptTokensDetails(
+                        cached_tokens=state.prefix_hit_length
+                    ),
+                    completion_tokens_details=CompletionTokensDetails(
+                        reasoning_tokens=state.reasoning_tokens
+                    ),
+                )
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/generator/batch_generate.py:313]: how realistic is it to extract this as a function of state & mlx_stats? seems like that's all this requires.

File: src/exo/worker/engines/mlx/generator/batch_generate.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+                        stats=stats,
+                        usage=usage,
+                    ),
+                )
+            )
+
+            if is_done:
+                del self._active_tasks[response.uid]
+            elif max_stop_len > 0 and len(state.accumulated_text) > max_stop_len:
+                state.accumulated_text = state.accumulated_text[-max_stop_len:]
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/generator/batch_generate.py]: if this is the only purpose of accumulated_text (scanning for stop sequences) could we rename it to something like `potential_stop_sequence_text`?

File: src/exo/worker/engines/mlx/utils_mlx.py [hunk 8]
```diff
@@ -750,3 +751,55 @@ def _parse_single_tool(text: str) -> dict[str, Any]:
...
+        return [
+            int.from_bytes(utf8_task_id[i : i + 1]) for i in range(len(utf8_task_id))
+        ]
+
+    def decode_task_id(encoded_task_id: list[int]) -> TaskId:
+        return TaskId(
+            bytes.decode(b"".join((x).to_bytes(length=1) for x in encoded_task_id))
+        )
+
+    uuid_byte_length = 36
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/utils_mlx.py:771]: this is a little spooky imo - can we assert the taskid is of appropriate length?

File: src/exo/worker/engines/mlx/utils_mlx.py [hunk 9]
```diff
@@ -750,3 +751,55 @@ def _parse_single_tool(text: str) -> dict[str, Any]:
...
+        .tolist(),
+    )
+    all_task_ids: list[list[TaskId]] = [
+        [decode_task_id(encoded_task_id) for encoded_task_id in rank_tasks[:count]]
+        for rank_tasks, count in zip(gathered, all_counts, strict=True)
+    ]
+
+    agreed_ids: set[TaskId] = set(all_task_ids[0])
+    for rank_tasks in all_task_ids[1:]:
+        agreed_ids &= set(rank_tasks)
```
Comment (Evanev7) [inline src/exo/worker/engines/mlx/utils_mlx.py]: do we prefer

agreed_ids = set[TaskId].intersection(*(set(tids) for tids in all_task_ids)?

File: src/exo/worker/runner/llm_inference/batch_generator.py [hunk 10]
```diff
@@ -294,53 +279,212 @@ def close(self) -> None:
...
+        agreed, different = mx_all_gather_tasks(self._maybe_queue, self.group)
+        self._queue.extend(task for task in self._maybe_queue if task in agreed)
+        self._maybe_queue = [task for task in self._maybe_queue if task in different]
+
+    def step(
+        self,
+    ) -> Iterable[
+        tuple[TaskId, GenerationResponse | ToolCallResponse | Cancelled | Finished]
+    ]:
+        self.agree_on_tasks()
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/batch_generator.py]: potential todo - should we merge submit() and step()? if we only ever actually submit() on step(), it could make sense to shrink the interface to step(new_tasks) -> list[Current Results] - though the connection between input and output is not obvious... i've talked myself out of it. potential future refactor could be to make submit(task) -> Return channel though.

File: src/exo/worker/runner/llm_inference/batch_generator.py [hunk 11]
```diff
@@ -294,53 +279,212 @@ def close(self) -> None:
...
+        int,
+        tuple[
+            TextGeneration,
+            GeneratorQueue[GenerationResponse],
+            Generator[GenerationResponse | ToolCallResponse | None],
+        ],
+    ] = field(default_factory=dict, init=False)
+
+    def __post_init__(self) -> None:
+        self._mlx_gen = ExoBatchGenerator(
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/batch_generator.py:334]: _mlx_gen -> _exo_gen

File: src/exo/worker/runner/llm_inference/batch_generator.py [hunk 12]
```diff
@@ -294,53 +279,212 @@ def close(self) -> None:
...
+    ) -> Iterable[
+        tuple[TaskId, GenerationResponse | ToolCallResponse | Cancelled | Finished]
+    ]:
+        self.agree_on_tasks()
+
+        # Submit any queued tasks to the engine
+        while self._queue and len(self._active_tasks) < EXO_MAX_CONCURRENT_REQUESTS:
+            task = self._queue.popleft()
+            try:
+                uid = self._build_generator(task)
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/batch_generator.py]: it is not clear that both _build_generator and _mlx_gen.submit **run prefill immediately** - i think we should keep behaviour as is but change the interface slightly

File: src/exo/worker/runner/llm_inference/batch_generator.py [hunk 13]
```diff
@@ -294,53 +279,212 @@ def close(self) -> None:
...
+                    self.tool_parser,
+                    self.tokenizer,
+                    type(self.model),
+                    self.model_id,
+                    task.task_params.tools,
+                )
+            self._active_tasks[uid] = (task, queue, output_generator)
+
+        if not self._mlx_gen.has_work:
+            return self._drain_cancellations()
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/batch_generator.py]: similarly "drain cancellations" imo implies removing a cancellation rather than removing it's corresponding task

File: src/exo/worker/runner/llm_inference/batch_generator.py [hunk 14]
```diff
@@ -294,53 +279,212 @@ def close(self) -> None:
...
+
+        results = self._mlx_gen.step()
+
+        output: list[
+            tuple[TaskId, GenerationResponse | ToolCallResponse | Cancelled | Finished]
+        ] = []
+        for uid, response in results:
+            if uid not in self._active_tasks:
+                # should we error here?
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/batch_generator.py:424]: ref comment and review comment above, i think at least a log is due here

File: src/exo/worker/runner/llm_inference/runner.py [hunk 15]
```diff
@@ -392,25 +377,57 @@ class Builder:
...
     inference_model: Model | None = None
     tokenizer: TokenizerWrapper | None = None
-    tool_parser: ToolParser | None = None
     group: mx.distributed.Group | None = None
-    kv_prefix_cache: KVPrefixCache | None = None
 
     def build(
         self,
     ) -> InferenceGenerator:
+        import os
```
Comment (Evanev7) [inline src/exo/worker/runner/llm_inference/runner.py]: no reason to import late here

**Review by Evanev7 — APPROVED** [review]
exciting!!! these are mostly stylistic changes with one or two minor correctness things we were probably doing wrong before anyway.

!!!!! continuous batching !!!!!

File: src/exo/worker/engines/mlx/generator/batch_generate.py (hunk 3 above)
Comment (rltakashige) [reply src/exo/worker/engines/mlx/generator/batch_generate.py]: I make assumptions that uids[0] is the real uid anyway, so will assert that.

File: src/exo/worker/engines/mlx/generator/batch_generate.py (hunk 6 above)
Comment (rltakashige) [reply src/exo/worker/engines/mlx/generator/batch_generate.py:313]: If we were to do something like this, I'd want to make both mlx_generate and this batch generate to be more similar in structure. But I think we can address this in in the future.
